### PR TITLE
Amb 1839 add cache to pds authenticator

### PR DESCRIPTION
**Pull request by artronics**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: lambda_code/src/pds_service.py [hunk 1]
```diff
@@ -22,22 +29,27 @@ def get_pds_secrets(self):
...
             "jti": str(uuid.uuid4())
         }
         return jwt.encode(claims, secret_object['private_key'], algorithm='RS512',
                           headers={"kid": secret_object['kid']})
 
     def get_access_token(self):
-        _jwt = self.create_jwt()
+        now = int(time.time())
+        cached = self.cache.get("pds_token")
+        if cached and cached["expires_at"] < now + self.expiry:
```
Comment (lee-richardson-nhs) [inline lambda_code/src/pds_service.py:48]: Shouldn't this be `now < cached["expires_at"]`? If it adds the token expiry duration then it's comparing against the expiry time if we made a new token now which will always be more (unless the config changes somehow so expiry duration is shorter).

File: lambda_code/tests/test_fhir_repository.py [hunk 2]
```diff
@@ -188,3 +249,75 @@ def test_delete_throws_error_when_response_can_not_be_handled(self):
...
+
+        # When
+        results = self.repository.find_immunizations("an-id", "a-code")
+
+        # Then
+        self.assertListEqual(results, [imms1, imms2])
+
+    def test_bad_response_from_dynamo(self):
+        """it should throw UnhandledResponse when the response from dynamodb can't be handled"""
+        bad_request = 400
```
Comment (lee-richardson-nhs) [inline lambda_code/tests/test_fhir_repository.py:314]: Should this be 500?

File: lambda_code/tests/test_cache.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import unittest
+
+sys.path.append(f"{os.path.dirname(os.path.abspath(__file__))}/../src")
+
+from cache import Cache
+
+
+class TestCache(unittest.TestCase):
+    def setUp(self):
+        self.cache = Cache(tempfile.gettempdir())
```
[the remark this answers is not in the document]
Comment (artronics) [reply lambda_code/tests/test_cache.py:14]: `setUp()` method is being called per test. So each test will start with a new instance. The setup for entire class during is called `setUpClass()`

File: lambda_code/tests/test_fhir_repository.py (hunk 2 above)
Comment (artronics) [reply lambda_code/tests/test_fhir_repository.py]: Any error response from dynamodb is an UnhandledResponse for us. So, in this case it doesn't make any difference.

File: lambda_code/tests/test_fhir_repository.py [hunk 4]
```diff
@@ -90,12 +109,54 @@ def test_create_should_catch_dynamo_error(self):
...
             # When
-            self.repository.create_immunization({"id": "an-id"})
+            self.repository.create_immunization(_make_an_immunization())
 
         # Then
         self.assertDictEqual(e.exception.response, response)
 
 
+class TestCreateImmunizationPatientIndex(unittest.TestCase):
+    """create_immunization should create a patient record with vaccine type"""
```
[the remark this answers is not in the document]
Comment (artronics) [reply lambda_code/tests/test_fhir_repository.py]: Sorry this is from outdated diff from old branch. The updated one has a test for diseaseType (which in this implementation is equivalent to diseaseCode). have a look and if you see it's missing then I'll add it to the ACs of AMB-1849 (amend search)

File: lambda_code/src/cache.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    def get(self, key: str) -> Optional[dict]:
+        return self.cache.get(key, None)
+
+    def delete(self, key: str):
+        if key not in self.cache:
+            return
+        del self.cache[key]
+
+    def _overwrite(self):
+        with open(self.cache_file.name, "w") as self.cache_file:
```
[the remark this answers is not in the document]
Comment (artronics) [reply lambda_code/src/cache.py:30]: Great point! Awesome. No,  AWS will create a filesystem per lambda environment. Here: https://repost.aws/questions/QUeWxoY-iYTBagcQznDPMBWg/lambda-concurrent-invocations-collision-using-same-tmp-yourfilename-csv

File: lambda_code/tests/test_cache.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    def test_cache_put(self):
+        """it should store cache in specified directory"""
```
[the remark this answers is not in the document]
Comment (artronics) [reply lambda_code/tests/test_cache.py:17]: It's a naming mistake. "directory" should be "key". I renamed it. The last test, test actual filesystem IO.

File: tests/test_crud_immunisation_api.py [hunk 7]
```diff
@@ -34,6 +34,7 @@ def create_a_deleted_imms_resource(imms_api: ImmunisationApi) -> dict:
         "login_form": {"username": "656005750104"},
     }
 )
+@pytest.mark.debug
```
Comment (lee-richardson-nhs) [inline tests/test_crud_immunisation_api.py:37]: Accidentally committed debug code?

**Review by lee-richardson-nhs — APPROVED** [review]

**Pull request merged**